libp2p/stream/lpstream.nim

```diff
@@ -109,6 +114,45 @@ proc readLine*(s: LPStream, limit = 0, sep = "\r\n"): Future[string] {.async, de
   except LPStreamIncompleteError, LPStreamReadError:
     discard # EOF, in which case we should return whatever we read so far..
 
+proc readVarint*(conn: LPStream): Future[uint64] {.async, gcsafe.} =
+  var
+    varint: uint64
+    length: int
+    buffer: array[10, byte]
+
+  for i in 0..<len(buffer):
+    await conn.readExactly(addr buffer[i], 1)
+    let res = PB.getUVarint(buffer.toOpenArray(0, i), length, varint)
+    if res == VarintStatus.Success:
+      return varint
+    if res != VarintStatus.Incomplete:
+      break
+  if true: # can't end with a raise apparently
+    raise (ref InvalidVarintError)(msg: "Cannot parse varint")
+
+proc readLp*(s: LPStream, maxSize: int): Future[seq[byte]] {.async, gcsafe.} =
```

Comment:
Not sure about this, it still doesn't make sense to pass a max size.

Reply:
Usually max size is set per protocol, not per read invocation.

Reply:
protocols have helpers for their own types though where they set this (`mplex.readMsg`) - you'll note that `readExactly`, `read`, etc all take a max size at this abstraction level

Reply:
potentially, `Connection` could have a helper without the max-len, assuming `Connection` is "heavy enough"

Reply:
consider also the multiple security problems we've been having with headers / negotiations - ie these typically have smaller max read sizes, so even within a protocol there will be differences

Reply:
https://github.com/status-im/nim-beacon-chain/blob/master/beacon_chain/eth2_network.nim#L265 - I'd like to use it here fore example with a custom chunk size

Reply:
well, now that I think about it, if we move it to lpstream, then it probably is OK to leave it with the max size, and have an overload without the max (or a global max).

Reply:
In fact, I'm thinking that the whole lpstream abstraction should go and connection should be the abstract base class that all other streams extend. This will fix all our closing issues which stem from the fact that we have to nest streams multiple times.